### ISSUE TITLE
Any way to prevent changes?

### ISSUE THREAD

**Issue by NoxLP** [gullerya/object-observer#120]
While using the observer I've seen that even if I throw an error in the `observe` callback, the changes still happen.

I need to prevent the changes to happen if certain conditions are wrong. Is there any way to achieve this?

**Comment by gullerya** [comment]
Hi,

The answer is no, there is no way to 'rollback' the change from within the observer and the fact that throwing error within the callback is not failing to update is by design - there are even tests dedicated to ensure that callback's error not only won't rollback the change but not even prevent other callbacks from being called.

I think that it would be wrong to implement such a functionality:
- observer is quite by it's very definition is not a mutative entity, but only observing
- in fact the way the implementation works the change as already happened, so we can't talk about 'prevent'
- callbacks order is not promised, so if some callback would prevent the change, but others has already been run assuming the change happened - total mess

The only thing I can suggest is to actually do a new change back to the prev value (you do have the new and the old values on the change object, as well as the essence of change) - this way you'll trigger a new change, which will be running a new callback/s invokation. Unless you are using and async flavor, this one will even be running only after all the callback/s of the current change done, so pretty much logically right.

**Comment by NoxLP** [comment]
It makes sense, I'll try your "rollback" suggestion.

Thanks for take your time and clarify it.

**Comment by DerKorb** [comment]
Would you be open to make this behaviour configurable? I am using only a single observer function and I am using it to validate permissions and track the changes. The setter not throwing when my observer throws on missing permissions makes everything around it super complicated for me. It would be really nice if this lib had a throwIfAnyObserverThrows setting or so.

**Comment by gullerya** [comment]
Hi, I've just reviewed the prev answer and the train of thoughts took me exactly to the same place.

I do not think it's a good approach to provide some ad hoc niche functionality, especially if it's not intuitive by definition. Observation design pattern is deliberately about side effect. Essentially you are talking about validation visitors, and I do not see it fit in this lib.

Of course, it could be made like: validators running before observers and any of them would throw the whole pipeline, but that's  just something else.

I'd suggest the same as above: do observe the changes, within observer you can do the validation and if prohibited - perform a rollback (you have the object, the prev value and the new one at hand, shouldn't be that complicated, especially if you know that it's the only observer running in the pipe).

**Comment by DerKorb** [comment]
In my case I provide a data object with your observer to identify changes to said object. My callback could be called PermissionsGuard. The object with observer is beeing used in some arbitrary reducers function which I as the author of the library do not know. If the author of the reducer function tries to modify the data but the user does not have the correct permission I would like the reducers to throw in place of the assignment. I could rollback after the reducer function has run or when it wants to use my library the next time, but this would not prevent some arbitrary call the reducers does inbetween. So for example if the reducer updates a users email adress, the reducer might then proceed to update this detail with an external auth service (eg. zitadel), then try to commit the transaction using my library and only then it would fail. Since I do not know, what the user of my library was doing inbetween I cant roll it back. For this reason it would be super usefull if the observer would throw directly in the setter, guranteeing nothing else would happen after that. If you dont want this kind of usage I can understand and forking or patching your code would work for me as well. I could see an argument for a guard that does not change data but only throws not beeing mutative and maybe providing this context would help you come to a similiar conclusion.

**Comment by gullerya** [comment]
I see, your case is not purely consumer but another lib/sdk that doesn't have the full logic control.

I'd still refrain from this change, since to introduce it properly is to substantially alter the behaviour, but I'll open another issue and will see for any interest. Please don't expect a quick resolution, if any.

**Comment by gullerya** [comment]
This is the issue I've mentioned: https://github.com/gullerya/object-observer/issues/149

**Comment by DerKorb** [comment]
no worries, for now i will just use pnpm patch feature to remove the try catch statement from handleCallbackSafe function which makes it work for me. Its just not a very nice solution so in the long term I would probably fork the code instead. What I would need is a setting whether exceptions in a callback are handled or not.  I dont think it would have to be a very large change, could be as simple as adding that as optional argument to observe function.

**Comment by DerKorb** [comment]
I have read your issue. I like that design. On validation failure, is your plan to just not apply the value or throw. Because for me it would be important the code throws in the location of the assignment. If it only does not apply the value and continues running, that would not work for me.

**Comment by gullerya** [comment]
I'd do contract by returning false to stop execution and prevent change, so
that would be the best practice, but there would be no try catch too, so
you, as a consumer, can throw and it'll do the same, just much more
expensive/brute :)
